### DISCUSSION TITLE
QEMU based podman for Windows

### DISCUSSION THREAD

**Discussion by imperialguy** [containers/podman#14796] [General]
Can you provide a QEMU based podman for windows instead of WSL2? The reason being WSL2 requires virtualization, hardware acceleration enabled etc. And that's quite a hard sell especially when it comes to Windows VMs running in corporate environments.

**Comment by rhatdan** [comment]
@n1hility @baude @gbraad WDYT?

**Reply by gbraad** [reply]
https://github.com/containers/podman/discussions/14796#discussioncomment-3064513

**Reply by gbraad** [reply]
@rhatdan from the comments of me, @n1hility and @afbjorklund we can conclude this is not a priority and more of an experiment. It is strongly suggested to use WHPX (Windows Hypervisor Platform)

**Reply by rhatdan** [reply]
SGTM

**Comment by gbraad** [comment]
Qemu is an option and is something we can look at, though the mechanism would differ from on Linux as the ignition config can not be set. IIRC we would also run into issues with the network options.

Ideally we would have to build our own variant to use WHPX/HAXM (the HyperV platform extensions, Intel's HW Accelerated Execution Manager) otherwise performance will suffer. One of the points you raise is to avoid paravirtualization (hypercalls), as you mention it is a hard sell to run VMs in corporate environments. This is not the experience from dealing for five years with Minikube/Minishift and CDK. Some corporations might restrict the use and prefer the user not having a 'root' equivalent inside the VM. 

Note: You can also try crc which uses HyperV instead to deploy Podman, however this also relies on native virtualization options.
Note2: I did experiment with a VERY minimal Podman image that was based on the prepped Fedora image for WSL2. Made it bootable with extlinux, etc. I am willing to continue testing with this, but not sure if this will become something standard, @n1hility WDYT?

![image](https://user-images.githubusercontent.com/1894/176892134-61b10551-cc0e-4758-8fd3-80e6f0ef455b.png)

Note3: I also ran this test without the `-accel whpx` option and the performance dropped significantly to about a half to a third.

**Comment by n1hility** [comment]
@imperialguy IMO you really need virt or the experience will be so poor that it's borderline unusable (as @gbraad indicates). I think if you are in a scenario where virt is inaccessible, then a remote container or a remote kube env is likely a better option than software emulation. Although, enabling local virt is fairly low overhead for an IT org. It's usually pre-enabled in hardware, so the change is really just group policy and feature enablement.

**Comment by afbjorklund** [comment]
I made a [proof-of-concept for Lima on Windows](https://github.com/lima-vm/lima/issues/909) (including Podman running on Fedora Cloud), using QEMU with WHPX... But it has the same requirements on hardware and OS license as other Hyper-V users, perhaps more since WSL2 runs also on Windows Home.

Podman example: https://github.com/lima-vm/lima/blob/master/examples/podman.yaml

Like was [shown above](https://github.com/containers/podman/discussions/14796#discussioncomment-3064513), the basic command used is: `qemu-system-x86_64 -accel whpx`

Agree that it would be better to use external servers, if local virtualization is not allowed by policy.
Doubly so if the host Windows already runs in a VM, then Linux could instead run in a second VM

**Reply by gbraad** [reply]
> since WSL2 runs also on Windows Home

Haven't looked into the latest specs, but I though Hyper-V only works on non-Home (non-core) installs and this is a dependency for WSL2. They have changed this? And if so, is that for Windows 11?

**Reply by afbjorklund** [reply]
I'm not a Windows person, but that is what they say:

https://docs.microsoft.com/en-us/windows/wsl/faq

> WSL 2 is available on all Desktop SKUs where WSL is available, including Windows 10 Home and Windows 11 Home.

Don't think they (Home) have a "full" Hyper-V license ?

**Comment by imperialguy** [comment]
Is QEMU something you are actively pursuing as an option or is WSL2 the official emulator for podman on Windows? 

Assuming virt is enabled, for metrics like performance, ease-of-use etc.., how would you rate CRC vs WSL2 vs QEMU on Windows? Which one of them is the best based on your experiments?

**Reply by afbjorklund** [reply]
You didn't reply on why a QEMU WHPX virtual machine would be better than the WSL2 system container (running in a virtual machine) ? They both require virtual machines and hardware acceleration, so right now the choice was more between "control" (like: bring your own kernel) and "licensing" (possibility to run under Windows Home) etc. Also one of the main benefits of QEMU as a platform is that it is both open source and cross platform, at least as used in lima and minikube. That was _my_ main reason, for it.

Don't have information about CRC or Podman Desktop platform decisions, so I guess you would have to compare their performance and ease-of-use yourself. But when it comes to technologies, both WSL2 and QEMU (with WHPX) are using Hyper-V "under the hood".

**Reply by afbjorklund** [reply]
Before QEMU, we used VirtualBox as the common virtualization platform. But it has issues, for instance with supporting ARM.

**Reply by imperialguy** [reply]
> You didn't reply on why a QEMU WHPX virtual machine would be better than the WSL2 system container (running in a virtual machine) ? 

I was initially trying to find out if QEMU would be an alternative to WSL2 without the virt (i.e., HPX stuff). But, based on the answers so far, it looks like they both need virt. So, I was wondering if WSL2 is the official emulator or you are pursuing QMEU or building something else completely new based on this comment:

> Qemu is an option and is something we can look at.....Ideally we would have to build our own variant to use WHPX/HAXM (the HyperV platform extensions, Intel's HW Accelerated Execution Manager) otherwise performance will suffer.

My main question was about virt vs non-virt - and it looks like it's been established that they all need virt irrespective. But, I see your point about licensed vs open-source. I like open-source stuff in general, although tbh, I don't have any strong opinion or bias on that aspect - either way is fine.

**Reply by afbjorklund** [reply]
You _can_ run qemu-system without hardware assistance, but it is very slow (like 10x) so only useful for small things and experiments.

For instance with lima it is used for running arm-on-intel or intel-on-arm. Generally one recommends using multi-arch or qemu-user.

Note:

"qemu-system" is the hypervisor, that runs the VM. "qemu-user" is a translation, for the application. (similar to Apple's Rosetta)

**Reply by afbjorklund** [reply]
The best is to run without virtualization, i.e. use Linux. The second best is to at least use the available hardware acceleration...

As far as I know, using WSL2 is the "official" way to run. For instance https://podman.io/getting-started/installation#windows

**Comment by afbjorklund** [comment]
Tracked as:
* https://github.com/containers/podman/issues/13006
